### PR TITLE
refactor: remove deferred function calls by using 'useMemo'

### PR DESCRIPTION
**Pull request by jfairley**
I'm looking to make changes according to https://github.com/Unleash/proxy-client-react/issues/22, but before I get started, I wanted to do this refactor in isolation. This way we can discuss it if anyone has questions.

In the former code, `useEffect` was being used more literally like `componentDidMount` for class components. While this works, since the `client` creation technically occurs on the 2nd pass of the context function, you see all the extra hoops to jump through with `deferCall`.

The spirit of this code is that the `client` is created up-front, exactly one time. `useMemo` accomplishes that, but this way, the `client` is defined on the first pass, allowing us to remove the `deferCall` code.

So to sum it up, nothing is different here, only better. I'll work on https://github.com/Unleash/proxy-client-react/issues/22 separately.

### CODE REVIEW

**Comment by jfairley** [comment]
Another note, I don't know how y'all do releases. It seems like y'all might follow https://www.conventionalcommits.org. If you're using `semantic-release`, I know `refactor` won't trigger a new release. If you prefer `fix:` I can change that.

**Comment by FredrikOseberg** [comment]
@jfairley Thanks for this! We actually just made a change to how this works to make it more compliant with SSR. I think it resolves the same use-case and has the added benefit of exposing the client for direct use. See #23

**Comment by jfairley** [comment]
Fixed by #23.
